### ISSUE TITLE
TF for Xeon Phi

### ISSUE THREAD

**Issue by oroojlooy** [tensorflow/tensorflow#8584]
Hi,

Does tensorflow support xeon phi? 

Thanks !
Afshin

**Comment by gunan** [comment]
Not yet. But we would be happy to accept this contribution into tensorflow.

**Comment by WorksWellWithOthers** [comment]
Hello. I just finished watching a webinar by Intel called, "TensorFlow* on Modern Intel® Architectures Webinar". It was unclear if TensorFlow now has support for Xeon phi cards. The webinar made me assume yes but now I'm under the impression that only Knights Landing is supported (which is not available?) and the Knights Corner is not supported. I am just trying to get a clear direction on how to setup my Knights Corner Xeon Phi card to use all of its core when training a model on TensorFlow.

**Comment by yaroslavvb** [comment]
I heard from Intel in January that they had their own fork of TensorFlow which was modified to support Xeon Phi and they were working on integrating it into the main repo, maybe someone familiar with the partnernship knows the status

**Comment by WorksWellWithOthers** [comment]
That's great news. Thank you.

**Comment by priyaarul** [comment]
Hi, 
Please check the below links, 

https://itpeernetwork.intel.com/tensorflow-intel-architecture-ai/
https://devmesh.intel.com/projects/intel-optimized-tensorflow1-0

Regards,
Priya

**Comment by WorksWellWithOthers** [comment]
@priyaarul 

Hi, so are you saying, "Yes, TensorFlow currently supports Knights Corner Xeon Phi cards."?

**Comment by cryptox31** [comment]
Hey,
Intel has released an article:
https://software.intel.com/en-us/articles/tensorflow-optimizations-on-modern-intel-architecture
https://software.intel.com/en-us/articles/intel-optimized-tensorflow-wheel-now-available#
which states that it does support tf on the xeon phi, but they haven't yet provided the documentation on how are they running the benchmarks. I have installed tf1.2 on the Xeon Phi Knight's Landing, but the models run too slow. @priyaarul @afshin67 @gunan @yaroslavvb

**Comment by WorksWellWithOthers** [comment]
@cryptox31 Did you find a way to run your models a bit faster? 
I'm trying to find out if my implementation for Knight's corner is incorrect or if I'm juicing it for all it's got. I should really just benchmark it against a GPU. If the Knight is slower I have my answer.

**Comment by mancoast** [comment]
Greetings @cryptox31 @WorksWellWithOthers 
Does the Intel Xeon Phi x100 PCI-E [Knight's corner] support Tensorflow?  Has anyone successfully executed the system on the 1st generation Xeon Phi?

Thanks,
coast

**Comment by gunan** [comment]
On TF side we still have not tested with avx512. So nothing from our end.

**Comment by mancoast** [comment]
greetings @gunan 
Please excuse the cross-link, but as a k1om vet, I strongly suggest that we consider implementation using https://github.com/edanor/umesimd .  This lib handles avx512 and the eol 512-bit instruction set of knights corner.  I would be more than willing to contribute, but require direction/instruction.

Thanks,
coast

**Comment by cryptox31** [comment]
@WorksWellWithOthers No, I haven't found a way yet but you can follow [this](https://software.intel.com/en-us/comment/1912006#comment-1912006) post on intel.

**Comment by priyaarul** [comment]
Hi All,please chk this-  https://software.intel.com/en-us/articles/tensorflow-optimizations-on-modern-intel-architecture- examples shown running inference on Intel® Xeon Phi™ processor 7250 (KNL)

**Comment by WorksWellWithOthers** [comment]
It's nice to have a TensorFlow wheel provided by Intel for their Xeon Phi cards.
Unfortunately, I'm getting warning messages that the binary was not compiled to use: SSE4.1, SSE4.2, AVX, AVX2, FMA.

Source:
https://software.intel.com/en-us/articles/intel-optimized-tensorflow-wheel-now-available#

# Python 2.7
pip install https://anaconda.org/intel/tensorflow/1.4.0/download/tensorflow-1.4.0-cp27-cp27mu-linux_x86_64.whl
 
# Python 3.5
pip install https://anaconda.org/intel/tensorflow/1.3.0/download/tensorflow-1.3.0-cp35-cp35m-linux_x86_64.whl
 
# Python 3.6
pip install https://anaconda.org/intel/tensorflow/1.4.0/download/tensorflow-1.4.0-cp36-cp36m-linux_x86_64.whl

Warning:
Python 2.7.5 (default, May  3 2017, 07:55:04) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-14)] on linux2                                                                            
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf 
>>> sess = tf.Session()                                                                                                                                      
2018-01-03 12:53:16.067051: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA                                                                                           

There is also a claim by someone stating that CPU with MKL is slower than just CPU support:
https://github.com/mind/wheels/issues/12

Are my times optimal or is something I can do to increase performance?

Sample:
sudo git clone https://github.com/tensorflow/models.git 
python models/tutorials/image/mnist/convolutional.py --batch_size=32

Step 0 (epoch 0.00), 3.2 ms                                                                                                                                  Minibatch loss: 8.334, learning rate: 0.010000                                                                                                               Minibatch error: 85.9%                                                                                                                                       Validation error: 84.6%                                                                                                                                      Step 100 (epoch 0.12), 26.2 ms                                                                                                                               Minibatch loss: 3.250, learning rate: 0.010000                                                                                                               Minibatch error: 6.2%                                                                                                                                        Validation error: 7.7%                                                                                                                                       Step 200 (epoch 0.23), 24.6 ms                                                                                                                               Minibatch loss: 3.359, learning rate: 0.010000                                                                                                               Minibatch error: 10.9%                                                                                                                                       Validation error: 4.2%                                                                                                                                       Step 300 (epoch 0.35), 24.4 ms                                                                                                                               Minibatch loss: 3.139, learning rate: 0.010000                                                                                                               Minibatch error: 3.1%                                                                                                                                        Validation error: 3.0%                                                                                                                                       Step 400 (epoch 0.47), 23.8 ms                                                                                                                               Minibatch loss: 3.197, learning rate: 0.010000                                                                                                               Minibatch error: 4.7%                                                                                                                                        Validation error: 2.6%                                                                                                                                       Step 500 (epoch 0.58), 24.0 ms                                                                                                                               Minibatch loss: 3.166, learning rate: 0.010000                                                                                                               Minibatch error: 6.2%                                                                                                                                        Validation error: 2.3%

**Comment by WorksWellWithOthers** [comment]
Doing the following gave me some improvement:

KMP_SETTINGS=1 KMP_BLOCKTIME=0 KMP_AFFINITY=granularity=fine,verbose,compact,1,0 OMP_NUM_THREADS=72 python models/tutorials/image/mnist/convolutional.py 

Step 100 (epoch 0.12), 20.9 ms                   <br />                                                                                                            Minibatch loss: 3.250, learning rate: 0.010000                     <br />                                                                                                             Minibatch error: 6.2%                          <br />                                                                                                                                 Validation error: 7.7%                            <br />                                                                                                                              Step 200 (epoch 0.23), 18.7 ms                        <br />                                                                                                                          Minibatch loss: 3.359, learning rate: 0.010000                           <br />                                                                                                       Minibatch error: 10.9%                              <br />                                                                                                                            Validation error: 4.2%                             <br />                                                                                                                             Step 300 (epoch 0.35), 18.7 ms                          <br />                                                                                                                        Minibatch loss: 3.139, learning rate: 0.010000                      <br />                                                                                                            Minibatch error: 3.1%                                           <br />                                                                                                                Validation error: 3.0%                                    <br />                                                                                                                      Step 400 (epoch 0.47), 18.6 ms                          <br />                                                                                                                        Minibatch loss: 3.197, learning rate: 0.010000                        <br />                                                                                                          Minibatch error: 4.7%                                               <br />                                                                                                            Validation error: 2.6%                                           <br />                                                                                                               Step 500 (epoch 0.58), 18.8 ms                             <br />                                                                                                                     Minibatch loss: 3.166, learning rate: 0.010000                       <br />                                                                                                           Minibatch error: 6.2%                                                <br />                                                                                                           Validation error: 2.3%                                           <br />                                                                                                               Step 600 (epoch 0.70), 18.8 ms                            <br />                                                                                                                      Minibatch loss: 3.121, learning rate: 0.010000                     <br />                                                                                                             Minibatch error: 3.1%                                           <br />                                                                                                                Validation error: 2.2%                                        <br />                                                                                                                  Step 700 (epoch 0.81), 18.7 ms                         <br />                                                                                                                         Minibatch loss: 2.970, learning rate: 0.010000                   <br />                                                                                                               Minibatch error: 1.6%                           <br />                                                                                                                                Validation error: 2.1%                                 <br />                                                                                                                         Step 800 (epoch 0.93), 18.6 ms                           <br />                                                                                                                       Minibatch loss: 3.040, learning rate: 0.010000                       <br />                                                                                                           Minibatch error: 4.7%                             <br />                                                                                                                              Validation error: 2.1%                            <br />                                                                                                                              Step 900 (epoch 1.05), 19.2 ms                           <br />                                                                                                                       Minibatch loss: 2.906, learning rate: 0.009500                 <br />                                                                                                                 Minibatch error: 1.6%                             <br />                                                                                                                              Validation error: 1.8%                         <br />                                                                                                                                 Step 1000 (epoch 1.16), 19.2 ms                  <br />                                                                                                                               Minibatch loss: 2.872, learning rate: 0.009500                     <br />                                                                                                             Minibatch error: 1.6%                                     <br />                                                                                                                      Validation error: 1.8%                                      <br />

**Comment by WorksWellWithOthers** [comment]
GPU benchmark claims (5ms?):
https://stackoverflow.com/questions/35703201/speed-benchmark-for-testing-tensorflow-install

**Comment by WorksWellWithOthers** [comment]
9ms: Geforce GTX 1060
![1060](https://user-images.githubusercontent.com/13571190/34545874-ba00c506-f0a4-11e7-87e2-d76283459bad.png)

**Comment by oroojlooy** [comment]
It seems that I need MPSS to manage the Xeon and I see that Intel just released the RedHat and Suse versions. I have Debian, can I install it? 
I found some instructions for converting the rpm packages into deb files. Do they work? 
Is there any suggestion?

**Comment by coldlionfanjin** [comment]
@WorksWellWithOthers I use tensorflow-mkl 1.8 on KNL, but my best result is about 65ms. Do you think this is caused by TF version?

**Comment by WorksWellWithOthers** [comment]
@coldlionfanjin 
Hi, it's been a while since I did these benchmarks and I no longer have access to the machine.

From my previous comments it seems this might help:
  KMP_SETTINGS=1 KMP_BLOCKTIME=0 KMP_AFFINITY=granularity=fine,verbose,compact,1,0 
  OMP_NUM_THREADS=72 python models/tutorials/image/mnist/convolutional.py

Best of luck!

P.S.
Most of my ML focus is now on game bots on consumer GPU cards.

**Comment by mohantym** [comment]
Hi @oroojlooy !
Is this issue still valid. Have you checked [installation guide ](https://www.intel.com/content/www/us/en/developer/articles/guide/optimization-for-tensorflow-installation-guide.html)from intel yet ? 
Thank you!
